fix inconsistency between puppeteer and sessionPool cookies, add tests

AndreyBykov opened the pull request:
For some cookies without an expiration date (session cookies), puppeteer would return `expires: -1` (chrome would show `Expires: Session` for these). Puppeteer methods would work completely fine while saving such cookies to session leads to some inconsistent behavior - either cookie would not be set to a session or it would not be then set to the page.

Fixes #922

AndreyBykov commented:
I am also thinking now - maybe it's better to remove the expiration date for these instead of setting the default max age per session pool?

mnmkng commented on src/session_pool/session.js:
```diff
@@ -305,7 +305,7 @@ export class Session {
      */
     _puppeteerCookieToTough(puppeteerCookie) {
-        const isExpiresValid = puppeteerCookie.expires && typeof puppeteerCookie.expires === 'number';
+        const isExpiresValid = puppeteerCookie.expires && typeof puppeteerCookie.expires === 'number' && puppeteerCookie.expires !== -1;
```
Would be better to check that it's non-negative, rather than exactly `-1`.

```js
const isExpiresValid = typeof puppeteerCookie.expires === 'number' && puppeteerCookie.expires > 0
```

AndreyBykov replied:
Also changing that, but in this case, I actually expect exactly `-1` as this is how puppeteer returns these cookies, if it's something else - it might be something completely different. On the other hand - other negative value would break it again, so yeah, better to have it more non-negative.

mnmkng commented on test/crawlers/puppeteer_crawler.test.js:
```diff
@@ -200,4 +200,38 @@ describe('PuppeteerCrawler', () => {
+
+        const puppeteerCrawler = new Apify.PuppeteerCrawler({
+            requestList,
+            useSessionPool: true,
```
please add `persistCookiesPerSession: true` to make it clear. Defaults can change.

AndreyBykov replied:
Adding it - but as in the issue conversation - this should not really affect anything, as we specifically save cookies to session in `createSessionFunction` -> so it should not matter whether it's specified or not...

mnmkng replied:
It does not matter now. But it could matter a year from now 😉

AndreyBykov commented:
1. Noticed that it was not possible to have `persistCookiesPerSession: false` and thus the bug from 0.22 was still there in the end (session would not be set to page with the above setting. Fixed.
2. `_toughCookieToPuppeteer` was returning the wrong `expires` value (value was in ms, while puppeteer/playwright both expect seconds and in the end, some weird values appeared like `+053122-01-20T01:44:13.000Z` which is `Jan 20 53122`). Fixed.

Noticed that the browser would still show the expiry date for some session cookies and won't treat them as session cookies, but could not find a way to make the browser recognize them correctly always, but I am still pretty sure we could leave the default cookie expiration date for these.

mnmkng commented on src/crawlers/browser_crawler.js:
```diff
@@ -251,7 +251,7 @@ class BrowserCrawler extends BasicCrawler {
             handlePageTimeoutSecs = 60,
             gotoFunction,
-            persistCookiesPerSession,
+            persistCookiesPerSession = true,
```
I think @petrpatek decided not to use the default here, because of the check on line 264. Without the default, the user can turn off session pool like that

```js
{
  useSessionPool: false,
}
```

with the default you need

```js
{
  useSessionPool: false,
  persistCookiesPerSession: false,
}
```

I think that's useful. You can fix it by making the check against the original options (I think).

```js
if (!useSessionPool && options.persistCookiesPerSession) {
```

Please also add comment to the check to explain why that is.

AndreyBykov replied:
Yeah, missed this part. I'll do it a bit differently - will remove the default here, and check `persistCookiesPerSession` against specifically `undefined` in `if (useSessionPool) {}`. If it's `undefined` - it's `true`, if not - it would be assigned with the value from options.

+ Would add `else` clause which would set `this.persistCookiesPerSession` to `false` (not to have it `undefined` as you mentioned in the next comment).

mnmkng commented on src/crawlers/browser_crawler.js:
```diff
@@ -285,7 +285,7 @@ class BrowserCrawler extends BasicCrawler {
 
         if (useSessionPool) {
-            this.persistCookiesPerSession = persistCookiesPerSession || true;
+            this.persistCookiesPerSession = persistCookiesPerSession;
```
Since we're changing the defaults, I would move this outside of the `if` clause. So that it's always a proper boolean and not sometimes `undefined`.

AndreyBykov replied:
Kept inside `if` clause, which would set to `true`, and also added `else` which would set it to `false`.

mnmkng commented on src/session_pool/session.js:
```diff
@@ -331,7 +331,7 @@ export class Session {
             name: toughCookie.key,
             value: toughCookie.value,
-            expires: new Date(toughCookie.expires).getTime(),
+            expires: new Date(toughCookie.expires).getTime() / 1000,
```
So this means that `toughCookie` saves `expires` in millisecond precision but Puppeteer expects it in seconds? Would be worth a comment.

AndreyBykov replied:
Exactly. Both playwright and puppeteer expect seconds, while tough cookie have actually Date string here (to create which we have to use millis).

mnmkng reviewed (changes requested):
Thanks, on a good track, just a few minute details. When reviewing I briefly checked `CheerioCrawler` and I'm not sure the handling there is consistent with the `BrowserCrawler`. Could you please check that we set the defaults correctly there and also that we use them correctly?

AndreyBykov commented:
Checked and updated `CheerioCrawler` too, should be consistent with `BrowserCrawler` now.

mnmkng commented on src/crawlers/browser_crawler.js:
```diff
@@ -285,7 +286,7 @@ class BrowserCrawler extends BasicCrawler {
 
         if (useSessionPool) {
-            this.persistCookiesPerSession = persistCookiesPerSession || true;
+            this.persistCookiesPerSession = persistCookiesPerSession !== undefined ? persistCookiesPerSession : true;
```
When `persistCookiesPerSession` is `undefined`, this will end up

```js
this.persistCookiesPerSession = undefined
```

AndreyBykov replied:
I am not quite following... If `persistCookiesPerSession` is `undefined` -> `this.persistCookiesPerSession` will be set to `true` when session pool is used, and will be set to `false` if `useSessionPool: false`.

Regarding the defaults - it actually becomes less readable - if I set the default to true (like I did before) - I have to perform the check with `options.persistCookiesPerSession` and then re-assign the value. If it's `false` by default - I would also have to re-assign the value for certain scenarios. Now it's always true or false in the end depending on actual usage of session pool (without the default value).

mnmkng replied:
Yup, sorry. Read it as `===`

mnmkng commented on src/crawlers/cheerio_crawler.js:
```diff
@@ -412,7 +413,12 @@ class CheerioCrawler extends BasicCrawler {
-        this.persistCookiesPerSession = persistCookiesPerSession;
+
+        if (this.useSessionPool) {
+            this.persistCookiesPerSession = persistCookiesPerSession !== undefined ? persistCookiesPerSession : true;
```
same here

AndreyBykov replied:
And it's done in exact same way here - if `persistCookiesPerSession` is not `undefined` - use the actual value, if it is `undefined` - set it to `true` when session pool is used, and set it to `false` when not. It's never `undefined`, or am I missing something?

mnmkng reviewed (changes requested):
I'd go with the defaults :)
